pysmt/solvers/btor.py

```diff
@@ -0,0 +1,347 @@
+
+        self.btor = boolector.Boolector()
+        self.btor.Set_opt("incremental", 1)
+        self.btor.Set_opt("model_gen", 1)
+        self.converter = BTORConverter(environment, self.btor)
+        self.mgr = environment.formula_manager
+        self.declarations = {}
+        return
+
+    @clear_pending_pop
+    def _reset_assertions(self):
+        raise NotImplementedError
+
+    @clear_pending_pop
+    def declare_variable(self, var):
+        raise NotImplementedError
+
+    @clear_pending_pop
+    def _add_assertion(self, formula, named=None):
+        self._assert_is_boolean(formula)
+        term = self.converter.convert(formula)
+        self.btor.Assert(term)
+
+    def get_model(self):
+        assignment = {}
+        for s, _ in self.converter.declared_vars.items():
+            assignment[s] = self.get_value(s)
+        return EagerModel(assignment=assignment, environment=self.environment)
+
+    @clear_pending_pop
+    def _solve(self, assumptions=None):
+        if assumptions is not None:
+            assumption = self.mgr.And(assumptions)
+            self._assert_is_boolean(assumption)
+            btor_assumption =  self.converter.convert(assumption)
+            self.btor.Assume(btor_assumption)
+
+        res = self.btor.Sat()
+        if res == self.btor.SAT:
+            return True
+        elif res == self.btor.UNSAT:
+            return False
+        else:
+            raise SolverReturnedUnknownResultError
+
+    def get_unsat_core(self):
+        raise NotImplementedError
+
+    @clear_pending_pop
+    def _push(self, levels=1):
+        # Boolector does not support push/pop.
+        # Incrementality could be simulated by keeping a stack
+        # of boolector instances.
+        # See self.btor.Clone()
+        raise NotImplementedError
+
+    @clear_pending_pop
+    def _pop(self, levels=1):
+        raise NotImplementedError
+
+    def print_model(self, name_filter=None):
+        for var in self.declarations:
+            if name_filter is None or not var.symbol_name().startswith(name_filter):
+                print("%s = %s" % (var.symbol_name(), self.get_value(var)))
+
+    def get_value(self, item):
+        self._assert_no_function_type(item)
+        return self.converter.back(item)
+
+    def exit(self):
+        if not self._destroyed:
+            self._destroyed = True
+            del self.btor
+
+
+class BTORConverter(Converter, DagWalker):
+
+    def __init__(self, environment, btor):
+        DagWalker.__init__(self, environment)
+        self.mgr = environment.formula_manager
+        self._get_type = environment.stc.get_type
+        self._back_memoization = {}
+        self._btor = btor
+        self.declared_vars = {}
+        return
+
+    @catch_conversion_error
+    def convert(self, formula):
+        return self.walk(formula)
+
+    def back(self, expr):
+        titem = self.convert(expr)
+        if expr.is_symbol() and  expr.symbol_type().is_bv_type():
```

Comment:
We should clarify when back conversion needs to be implemented. I am moving the code that was here to ```get_value```. I am not sure that a proper implementation of ```back``` is needed, at this point.